### PR TITLE
[RFC] Use JWT to configure the front controller

### PR DESCRIPTION
**Pull request by aschempp**
Here's an early draft of @terminal42's approach to new entry points.

- [x] ~~instead of having multiple entry points (`app.php`, `app_dev.php`), we want to have only one entry point (`index.php` as in Symfony 4).~~ Not possible, we still need a separate entry point to have a different URL to bypass the cache.
- [x] to know the user without needing to boot the kernel, a JWT cookie can be used

the JWT can tell
 - [x] ~whether to bypass the cache (if JWT token says the user is logged into the backend)~ If the user is allowed to use the front controller that bypasses the cache
 - [x] to enable or disable debug mode from the backend
 - [x] ~to enable or disable preview mode (without the iframe)~ Preview is determined by the user token, and we shouldn't change that for BC

this allows us to
 - [x] ~remove dependency on `terminal42/header-replay-bundle` to bypass cache if a user is logged in~ see #365
 - [x] ~remove dependency on session/user/ip etc. to bypass (lexik) maintenance mode~
 - [x] allow to use the debug mode without needing to activate it in the Contao Manager
 - probably a lot more 🎉 

regarding the JWT
 - [x] because we cannot use the kernel secret (the kernel/container is not bootet yet), the secret must be automatically generated and stored in a temporary file (e.g. `/var/cache/jwt-secret`)
 - [x] the JWT can only be signed using this secret, we can't encrypt it because we cannot rely on OpenSSL to generate a valid certificate on-the-fly

### CODE REVIEW

File: core-bundle/src/Resources/contao/controllers/BackendMain.php [hunk 1]
```diff
@@ -249,6 +285,10 @@ protected function output()
 		$this->Template->preview = $GLOBALS['TL_LANG']['MSC']['fePreview'];
 		$this->Template->previewTitle = StringUtil::specialchars($GLOBALS['TL_LANG']['MSC']['fePreviewTitle']);
 		$this->Template->profile = $GLOBALS['TL_LANG']['MSC']['profile'];
+		$this->Template->canDebug = $this->User->isAdmin;
+		$this->Template->isDebug = $container->get('kernel')->isDebug();
+		$this->Template->debug = $container->get('kernel')->isDebug() ? $GLOBALS['TL_LANG']['MSC']['disableDebugMode'] : $GLOBALS['TL_LANG']['MSC']['enableDebugMode'];
+		$this->Template->referer = base64_encode($container->get('request_stack')->getCurrentRequest()->getQueryString());
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/controllers/BackendMain.php]: Null pointer exception may occur here

File: core-bundle/src/Resources/contao/library/Contao/Template.php [hunk 2]
```diff
@@ -319,6 +319,24 @@ public function route($strName, $arrParams=array())
 		return ampersand($strUrl);
 	}
 
+	public function previewRoute($strName, $arrParams=array())
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/library/Contao/Template.php:330]: Are you ok with renaming this to just `preview()`?

File: manager-bundle/src/Resources/skeleton/web/app.php [hunk 3]
```diff
@@ -1,60 +1,4 @@
...
-// Enable the Symfony reverse proxy if request has no surrogate capability
-if (null !== $cache->getSurrogate() && !$cache->getSurrogate()->hasSurrogateCapability($request)) {
-    $kernel = $cache;
-}
-
-// Handle the request
-$response = $kernel->handle($request);
-$response->send();
-$kernel->terminate($request, $response);
+// This is just a BC layer
```
Comment (leofeyer) [inline manager-bundle/src/Resources/skeleton/web/app.php:3]: Really not sure about this…

File: manager-bundle/src/Resources/skeleton/web/preview.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+$loader = require __DIR__.'/../vendor/autoload.php';
+
+$request = Request::createFromGlobals();
+$jwtManager = new JwtManager(\dirname(__DIR__));
+
+try {
+    $jwt = $jwtManager->parseRequest($request);
+} catch (ResponseException $e) {
+    $e->getResponse()->send();
+    exit;
```
Comment (leofeyer) [inline manager-bundle/src/Resources/skeleton/web/preview.php:38]: Hm…

**Review by leofeyer — CHANGES_REQUESTED** [review]

File: core-bundle/src/Resources/contao/library/Contao/Template.php (hunk 2 above)
Comment (aschempp) [reply core-bundle/src/Resources/contao/library/Contao/Template.php:330]: No

File: core-bundle/tests/Functional/app/config/security.yml [hunk 5]
```diff
@@ -16,7 +16,7 @@ security:
             security: false
 
         contao_install:
-            pattern: ^/contao/install
+            pattern: ^/contao/install$
```
Comment (aschempp) [inline core-bundle/tests/Functional/app/config/security.yml]: This is not related to the pull request

File: core-bundle/tests/Functional/app/config/security.yml [hunk 6]
```diff
@@ -79,4 +79,5 @@ security:
 
     access_control:
         - { path: ^/contao/login$, roles: IS_AUTHENTICATED_ANONYMOUSLY }
+        - { path: ^/contao/logout$, roles: IS_AUTHENTICATED_ANONYMOUSLY }
```
Comment (aschempp) [inline core-bundle/tests/Functional/app/config/security.yml]: This does not seem to be related to the PR

File: core-bundle/tests/Functional/app/config/security.yml (hunk 5 above)
Comment (leofeyer) [reply core-bundle/tests/Functional/app/config/security.yml]: True. I will revert.

**Comment by leofeyer** [comment]
Thank you @aschempp.
